### PR TITLE
Add support for wasi-http

### PR DESCRIPTION
**Pull request by brendandburns**
@achille-roussel as discussed in https://github.com/tetratelabs/wazero/pull/1519 this adds wasi-http support.

This code is very incomplete and hacky, but it is working for limited use cases.

If this integration looks correct-ish, I will polish this up, add tests etc.

Thanks!

### CODE REVIEW

File: cmd/wasirun/main.go
Comment (achille-roussel) [inline cmd/wasirun/main.go:61]: I am tempted to suggest we enable by default, what do you think @chriso ?

File: imports/wasi_http/default_http/http.go
Comment (achille-roussel) [inline imports/wasi_http/default_http/http.go:9]: Is there a plan to migrate this module name to `wasi_http` (or `wasi-http` maybe)?

Keeping a correspondence between the WebAssembly host module and the Go package helps with discovery and reducing the cognitive overhead.

If this module is expected to remain called as-is, I would suggest to change the Go package name to `default_outgoing_http`, what do you think?

File: imports/wasi_http/default_http/request.go
Comment (achille-roussel) [inline imports/wasi_http/default_http/request.go:32]: Is there a way to report more descriptive error codes to the guest?

File: imports/wasi_http/wasi_streams/streams.go
Comment (achille-roussel) [inline imports/wasi_http/wasi_streams/streams.go:1]: Similar comment on naming consistency between the module and Go package here. How about renaming to `streams` since that's the name of the host module. It's a bit generic but so is `types` and the namespacing under `wasi_http` still clarifies what this is about I think.

File: imports/wasi_http/wasi_streams/streams.go
Comment (achille-roussel) [inline imports/wasi_http/wasi_streams/streams.go:20]: Let's change those to camel case to follow Go naming convensions?

**Review by achille-roussel — COMMENTED** [review]
Thanks for the contribution!

It's great to have the change somewhat self-contained in the `imports/wasi_http`, a few things I'd love to see added:

- I know we didn't set a great precedent with `imports/wasi_snapshot_preview1`, tho adding a README to `imports/wasi_http` would be great: I'd love to have things like a short description of what the module does, maybe a code example of how to use it, and links that will be relevant to help us maintain the code (e.g. locations where this module is implemented for other runtimes, related discussions, specification drafts, etc...)

- Having a guest program to test the module, we keep a few of those in https://github.com/stealthrocket/wasi-go/tree/main/testdata; this would be extremely important to ensure that we don't introduce regressions when we will make updates to this code. I know WASI makes constructing client/server applications challenging, this gist may be a good starting point to put together a test program using standard WASI features https://gist.github.com/chriso/6c71e968ef1002981a6ff46ceaa39ee3

File: cmd/wasirun/main.go
Comment (chriso) [reply cmd/wasirun/main.go:61]: One of the nice things about the sockets extensions is that wasi-go can autodetect which one you're using in https://github.com/stealthrocket/wasi-go/blob/main/imports/extensions.go. It'd be nice to do the same thing with this host module. If the module requires the HTTP host imports, we'd automatically set up the right host module. If new versions of `wasi-http` come out in future then ideally we'd be able to detect which one is in use and set it up automatically.

How about something like `-http <auto|none|v1(|v2...)>` which defaults to `auto`?

File: imports/wasi_http/default_http/request.go
Comment (brendandburns) [reply imports/wasi_http/default_http/request.go:32]: Unfortunately, there's really not. The method is defined here:

https://github.com/WebAssembly/wasi-http/blob/main/wit/types.wit#L86

In theory we could reserve some numbers in this implementation, but it isn't spec'd out in the current interface.

Comment (brendandburns) [reply imports/wasi_http/default_http/request.go:32]: I filed this issue on the spec: https://github.com/WebAssembly/wasi-http/issues/44

File: imports/wasi_http/wasi_streams/streams.go
Comment (brendandburns) [reply imports/wasi_http/wasi_streams/streams.go]: done. (and elsewhere)

**Comment by brendandburns** [comment]
Ok, since this looks good, I will push it forward to something that might actually be mergeable (will likely take a few days because of real life)

Will also add the README.md and a test.

File: imports/wasi_http/wasi_streams/streams.go
Comment (brendandburns) [reply imports/wasi_http/wasi_streams/streams.go]: renamed to `streams` some of this will clean up a little bit once wasi preview2 is implemented, since streams are part of that.

File: cmd/wasirun/main.go
Comment (brendandburns) [reply cmd/wasirun/main.go]: Done.

**Comment by brendandburns** [comment]
@achille-roussel this is now ready for review. I added a simple integration test. The coverage isn't great, but I plan to expand it in future PRs. Let me know if you want me to expand it before this merges.

Thanks!

File: testdata/c/http/proxy_component_type.o
Comment (achille-roussel) [inline testdata/c/http/proxy_component_type.o:1]: This file looks like an intermediary compilation object, do you know if it needs to be checked in?

**Review by achille-roussel — APPROVED** [review]
This is looking good!

I left a question about the `.o` file that was checked in, will merge when you get back to us @brendandburns

Thanks for the contribution 🙌

File: testdata/c/hello_world.wasm
Comment (achille-roussel) [inline testdata/c/hello_world.wasm:1]: Do you happen to have the sources for this file?

If not, I imagine we can reproduce it pretty easily, I can handle that post-merge if needed.

File: testdata/c/http/proxy_component_type.o
Comment (brendandburns) [reply testdata/c/http/proxy_component_type.o:1]: This is actually a by-product of running `wasi-bindgen`

If you'd prefer, I can set up the Makefile to download that binary and re-generate those artifacts if you want.

File: testdata/c/hello_world.wasm
Comment (brendandburns) [reply testdata/c/hello_world.wasm:1]: This file was already checked in:
https://github.com/stealthrocket/wasi-go/blob/main/testdata/c/hello_world.wasm

I'm not sure if that is on purpose. I suspect that it changed because I was using a different version of the `clang` compiler than when that test was checked in when I re-ran the tests.

If you want, I can remove it or revert my changes, let me know.

**Comment by brendandburns** [comment]
Thanks for the review. I can squash commits if you want before it merges (or you can squash on merge)

File: testdata/c/hello_world.wasm
Comment (achille-roussel) [reply testdata/c/hello_world.wasm:1]: Ah all good, this isn't a big deal if it's just a compiler version thing, I missed that we had the source checked in already 👍

**Review by achille-roussel — APPROVED** [review]
🚀 

I'll squash on merge 👍

**Comment by achille-roussel** [comment]
I tagged a `v0.7.0` release which includes these changes.

**Pull request merged**
